### PR TITLE
Add mailchimp newsletter subscription

### PR DESCRIPTION
**Pull request by nuno-aac**
This pr adds the new mailchimp subscription, the captcha is now invisible.

[Task.](https://www.notion.so/untile/Create-a-duplication-of-the-newsletter-f08911f2c6ae49e69a039916d4f36b37?pvs=4)
[Preview.](https://feature-update-newsletter-ma.starknet-websites.pages.dev/en/roadmap)

### CODE REVIEW

File: workspaces/website/src/api/index.ts [hunk 1]
```diff
@@ -48,3 +53,70 @@ apiRouter.get(
...
+      const captchaResult = await captchaResponse.json() as { success: boolean };
+
+      if (!captchaResult.success) {
+        return corsify(error(
+          422,
+          { title: 'Invalid Captcha' }
+        ));
+      }
+
+      const mailchimpUrl = `https://${mailchimpServer}.api.mailchimp.com/3.0/lists/${mailchimpList}/members`;
```
Comment (rafazsh) [inline workspaces/website/src/api/index.ts]: remove the `workspaces/website/src/api/constants.ts` file and create an env variable like `VITE_MAILCHIMP_NEWSLETTER_URL` where
```tsx
VITE_MAILCHIMP_NEWSLETTER_URL=https://us12.api.mailchimp.com/3.0/lists/a7c9440979/members
```

File: workspaces/website/src/api/index.ts [hunk 2]
```diff
@@ -48,3 +53,70 @@ apiRouter.get(
...
+
+      if(mailchimpResponse.ok !== true) {
+        return corsify(error(
+          mailchimpResponse.status,
+          mailchimpResult
+        ));
+      }
+
+      return corsify(json({
+        message: 'Successfully subscribed to newsletter!',
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/api/index.ts:110]: shouldn't this translation come from crowdin?

File: workspaces/website/src/api/index.ts [hunk 3]
```diff
@@ -48,3 +53,70 @@ apiRouter.get(
...
+        ...mailchimpResult
+      }));
+    } catch (err) {
+      return corsify(error(
+        (err as any)?.status ?? 500,
+        (err as any)?.response?.text ? JSON.parse((err as any)?.response?.text) : {
+          error: 'Internal Server Error'
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/api/index.ts:117]: same as above

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 4]
```diff
@@ -1,32 +1,185 @@
...
+    setIsSuccess(false);
   };
 
+  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    try {
+      setIsLoading(true)
+      const token = captchaRef.current?.getResponse();
+      await axios.post(`/api/newsletter-subscribe?${qs.stringify({
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:62]: break line

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 5]
```diff
@@ -1,32 +1,185 @@
...
   const handleClose = () => {
     setIsOpen(false);
+    setIsSuccess(false);
   };
 
+  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    try {
+      setIsLoading(true)
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx]: move this out of the `try`

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 6]
```diff
@@ -1,32 +1,185 @@
...
+ * `RoadmapSubscribeForm` component.
+ */
+
 function RoadmapSubscribeForm({
+  env,
   isOpen,
   setIsOpen,
 }: RoadmapSubscribeFormProps) {
+  const toast = useToast();
+  const [isLoading, setIsLoading] = useState(false);
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx]: why not a unique state where
```tsx
const [formState, setFormState] = useState<'submitting' | 'success' | 'error' | null>(null);
```

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 7]
```diff
@@ -1,32 +1,185 @@
...
+      >
         <ModalCloseButton />
-        <iframe
-          src="https://cdn.forms-content.sg-form.com/2fb14a4a-f24d-11ed-97a1-2255cc459392"
-          style={{ width: "100%", height: "420px" }}
-          frameBorder="0"
-        />
+
+        <Container 
+          padding={'60px 40px 40px !important'}
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 8]
```diff
@@ -1,32 +1,185 @@
...
+              >
+                Starknet Dev News ✨🗞️
+              </Heading>
+
+              <Text 
+                variant={'cardBody'}
+                color={'heading-navy-fg'}
+                paddingBottom={'32px'}
+              >
+                Receive notifications on Starknet version updates and network status.
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:130]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 9]
```diff
@@ -1,32 +1,185 @@
...
+                <FormControl isRequired paddingBottom={'20px'}>
+                  <FormLabel fontWeight={'700'}>
+                    Email
+                  </FormLabel>
+
+                  <Input
+                    type={'email'}
+                    name={'email'}
+                    id={'email'}
+                    placeholder={'Please enter your email'}
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:143]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 10]
```diff
@@ -1,32 +1,185 @@
...
+                />
+
+                <Button
+                  fontSize={'14px'}
+                  isLoading={isLoading}
+                  padding={'6px 16px'}
+                  type={'submit'}
+                  variant={'solid'}
+                >
+                  Submit
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:162]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 11]
```diff
@@ -1,32 +1,185 @@
...
+                </Button>
+              </form>
+            </>
+          ) : (
+            <div style={{ textAlign: 'center' }}>
+              <Text 
+                variant={'cardBody'}
+                color={'heading-navy-fg'}
+              >
+                Thank you and may the STARK be with you ✨🗞️
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:172]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 12]
```diff
@@ -1,32 +1,185 @@
...
+              </Text>
+            </div>
+          )}
+        </Container>
+
+        {/*<iframe
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx]: 👀

File: workspaces/website/src/api/index.ts [hunk 13]
```diff
@@ -48,3 +53,70 @@ apiRouter.get(
...
+/**
+ * Newsletter subscribe api route
+ */
+
+apiRouter.post(
+  '/newsletter-subscribe',
+  async (req, env: PAGES_VARS) => {
+    try {
+      const formData = new FormData();
+      formData.append('secret', env.CLOUDFLARE_RECAPTCHA_KEY);
```
Comment (rafazsh) [inline workspaces/website/src/api/index.ts:66]: add empty line between

File: workspaces/website/src/api/index.ts [hunk 14]
```diff
@@ -1,4 +1,9 @@
+/**
+ * Module dependencies
```
Comment (rafazsh) [inline workspaces/website/src/api/index.ts]: missing period in doc blocks

File: workspaces/website/src/api/index.ts [hunk 15]
```diff
@@ -48,3 +53,70 @@ apiRouter.get(
     );
   }
 );
+
+/**
+ * Newsletter subscribe api route
```
Comment (rafazsh) [inline workspaces/website/src/api/index.ts]: missing period in doc blocks

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 16]
```diff
@@ -1,32 +1,185 @@
...
+        token,
+      })}`)
+
+      setIsLoading(false)
+      setIsSuccess(true)
+    } catch (error: any) {
+      setIsLoading(false)
+      if(error.response.data?.title === 'Invalid Captcha') {
+        toast({
+          description: 'We\'re having trouble verifying you\'re a human. Please try again.',
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:73]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 17]
```diff
@@ -1,32 +1,185 @@
...
+          isClosable: true,
+          status: 'error',
+        });
+
+        return;
+      }
+
+      if(error.response.data?.title === 'Member Exists') {
+        toast({
+          description: 'You are already subscribed to the newsletter.',
```
Comment (rafaelcruzazevedo) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx:84]: 👀

File: workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx [hunk 18]
```diff
@@ -1,32 +1,185 @@
...
+      })}`)
+
+      setIsLoading(false)
+      setIsSuccess(true)
+    } catch (error: any) {
+      setIsLoading(false)
+      if(error.response.data?.title === 'Invalid Captcha') {
+        toast({
+          description: 'We\'re having trouble verifying you\'re a human. Please try again.',
+          duration: 1500,
```
Comment (rafazsh) [inline workspaces/website/src/pages/(components)/roadmap/RoadmapSubscribeForm.tsx]: maybe you can create a constant `toastConfig`
```tsx
const toastConfig = {
  duration: 1500,
  isClosable: true,
  status: 'error'
};

toast({
  description: 'foobar',
  ...toastConfig
});
```
and reuse in others

**Review by rafazsh — CHANGES_REQUESTED** [review]

File: workspaces/website/src/api/index.ts [hunk 19]
```diff
@@ -48,3 +52,69 @@ apiRouter.get(
...
+
+      const captchaResult = await captchaResponse.json() as { success: boolean };
+
+      if (!captchaResult.success) {
+        return corsify(error(
+          422,
+          { title: 'Invalid Captcha' }
+        ));
+      }
+      const mailchimpResponse = await fetch(
```
Comment (rafazsh) [inline workspaces/website/src/api/index.ts:84]: add empty line

**Review by rafazsh — CHANGES_REQUESTED** [review]

**Review by rafazsh — APPROVED** [review]

**Pull request merged**